File: python3/simplify_path.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def simplifyPath(self, path: str) -> str:
        stack = []
        splitted_path = path.split("/")

        for directory in splitted_path:
            if directory == "" or directory == ".":
                continue

            if directory == "..": 
                if len(stack) > 0:
                    stack.pop()
            else:
                stack.append(directory)

        return "/" + "/".join(stack)

File: golang/simplify_path.go
package golang

// Space	: O(n)
// Time		: O(n)

import (
	"strings"
)

func simplifyPath(path string) string {
	s := strings.Split(path, "/")
	newPath := []string{}

	for i := 0; i < len(s); i++ {
		if s[i] == "" || s[i] == "." {
			continue
		} else if s[i] == ".." {
			if len(newPath) > 0 {
				newPath = newPath[:len(newPath)-1]
			}
		} else {
			newPath = append(newPath, s[i])
		}
	}

	ans := ""
	for _, v := range newPath {
		ans += "/" + v
	}

	if ans == "" {
		ans += "/"
	}

	return ans
}
